internal/check/check.go

```diff
@@ -0,0 +1,847 @@
+		attribute.String("tuple_key", tuple.TupleKeyWithConditionToString(req.GetTupleKey())),
+	))
+	defer span.End()
+
+	node, ok := r.model.GetNodeByID(tuple.ToObjectRelationString(req.GetObjectType(), req.GetTupleKey().GetRelation()))
+	if !ok {
+		// this should never happen as the request is already validated before
+		return nil, ErrPanicRequest
+	}
+
+	// GetUserType returns the user type if the request in not an object relation otherwise the usertyperelation
+	_, ok = r.model.GetNodeWeight(node, req.GetUserType())
+	if !ok {
+		// If the user type is not reachable from the object type and relation, we can immediately return false.
+		return &Response{Allowed: false}, nil
+	}
+
+	// Check if the user is a wildcard and if the node does not have a path to the user type wildcard then return false
+	if req.IsTypedWildcard() && !slices.Contains(node.GetWildcards(), req.GetUserType()) {
+		return &Response{Allowed: false}, nil
+	}
+
+	return r.ResolveUnion(ctx, req, node, nil)
+}
+
+func (r *Resolver) isCached(consistency openfgav1.ConsistencyPreference, key string) (*Response, bool) {
+	if consistency == openfgav1.ConsistencyPreference_HIGHER_CONSISTENCY || r.lastCacheInvalidationTime.IsZero() {
+		return nil, false
+	}
+	v := r.cache.Get(key)
+	if v == nil {
+		return nil, false
+	}
+	res, ok := v.(*ResponseCacheEntry)
+	if !ok {
+		return nil, false
+	}
+	if !res.LastModified.After(r.lastCacheInvalidationTime) {
+		return nil, false
+	}
+	return res.Res, true
+}
+
+func buildEdgeCacheKey(modelID string, req *Request, edge *authzGraph.WeightedAuthorizationModelEdge) string {
+	keyBuilder := &strings.Builder{}
+	keyBuilder.WriteString(CacheKeyPrefix)
+	keyBuilder.WriteString(modelID)
+	keyBuilder.WriteString(DELIMITER)
+	keyBuilder.WriteString(req.GetTupleKey().GetObject())
+	keyBuilder.WriteString(DELIMITER)
+	keyBuilder.WriteString(req.GetTupleKey().GetUser())
+	keyBuilder.WriteString(DELIMITER)
+	keyBuilder.WriteString(edge.GetRelationDefinition())
+	keyBuilder.WriteString(req.GetInvariantCacheKey())
+	return keyBuilder.String()
+}
+
+func (r *Resolver) ResolveUnionEdges(ctx context.Context, req *Request, edges []*authzGraph.WeightedAuthorizationModelEdge, visited *sync.Map) (*Response, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	if res, ok := r.isCached(req.GetConsistency(), req.GetCacheKey()); ok {
+		return res, nil
+	}
+
+	out := make(chan ResponseMsg, len(edges))
+	var pool errgroup.Group
+	pool.SetLimit(r.concurrencyLimit)
+	defer func() {
+		_ = pool.Wait()
+		close(out)
+	}()
+
+	ids := make([]string, 0, len(edges))
+	for _, edge := range edges {
+		id := buildEdgeCacheKey(r.model.GetModelID(), req, edge)
+		ids = append(ids, id)
+		if res, ok := r.isCached(req.GetConsistency(), id); ok {
+			concurrency.TrySendThroughChannel(ctx, ResponseMsg{ID: id, Res: res}, out)
+			continue
+		}
+		pool.Go(func() error {
+			res, err := r.ResolveEdge(ctx, req, edge, visited)
```

Comment:
Want to double confirm here - visited can be nil.  In this case, each Edge will have their own version of sync map.  
If visited != nil, each Edge will share the sync map.